test/tests/functional/pbs_apbasil_inventory.py

```diff
@@ -0,0 +1,223 @@
+#
+# Use of Altair’s trademarks, including but not limited to "PBS™",
+# "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+# trademark licensing policies.
+
+from tests.functional import *
+from string import Template
+import os
+import tempfile
+import xml.etree.ElementTree as ET
+
+
+class TestBasilInventory(TestFunctional):
+    """
+    This test suite is for testing the support for BASIL 1.4 inventory
+    query.Test if query is made with correct BASIL version, and that
+    vnodes are getting created as per the query response.
+    """
+    basil_version = ['1.4', '1.3', '1.2', '1.1']
+
+    def restartPBS(self):
+        try:
+            svcs = PBSInitServices()
+            svcs.restart()
+        except PbsInitServicesError, e:
+            self.logger.error("PBS restart failed: \n" + e.msg)
+            self.assertTrue(e.rv)
+
+    def reset_nodes(self, hostA):
+        # Remove all nodes
+        rv = self.server.manager(MGR_CMD_DELETE, NODE, None, "")
+        self.assertEqual(rv, 0)
+
+        # Restart PBS
+        self.restartPBS()
+
+        # Create node
+        rv = self.server.manager(MGR_CMD_CREATE, NODE, None, hostA)
+        self.assertEqual(rv, 0)
+
+        # Wait for 3 seconds for changes to take effect
+        time.sleep(3)
+
+    def setUp(self):
+        momA = self.mom
+        mom_config = momA.parse_config()
+        if '$alps_client' not in mom_config:
+            self.skipTest("alps_client not set in mom config.")
+
+        if '$vnode_per_numa_node' in mom_config:
+            momA.unset_mom_config('$vnode_per_numa_node', False)
+
+        # check if required BASIL version available on the machine.
+        xml_out = self.query_alps(self.basil_version[0], 'QUERY', 'ENGINE')
+        xml_tree = ET.parse(xml_out)
+        os.remove(xml_out)
+
+        response = xml_tree.find(".//ResponseData")
+        status = response.attrib['status']
+
+        if status != "SUCCESS":
+            self.skipTest("Basil version %s not available on the platform" %
+                          (self.basil_version[0]))
+
+        # Reset nodes
+        self.reset_nodes(momA.shortname)
+
+    def query_alps(self, ver, method, qtype):
+        """
+        Send a query to ALPS of a certain type and return the xml output file.
+        """
+        basil_protocol = 'protocol="%s"' % (ver)
+        basil_method = 'method="%s"' % (method)
+        basil_qtype = 'type="%s"' % (qtype)
+        queryt = Template('<BasilRequest $ver $method $qtype/>\n')
+        query = queryt.substitute(ver=basil_protocol,
+                                  method=basil_method, qtype=basil_qtype)
+        mom_config = self.mom.parse_config()
+        alps_client = mom_config['$alps_client']
+
+        (fd, fn) = tempfile.mkstemp()
+        os.write(fd, query)
+        os.close(fd)
+        (fd, xout) = tempfile.mkstemp()
+        os.close(fd)
+        os.system("%s < %s > %s" % (alps_client, fn, xout))
```

Comment:
use du.run_cmd instead of system

Reply:
Hi Arun,
   I spend a lot of time to make du.run_cmd work for above scenario but it's not working. Redirection is failing for the above command and command is nor reading the input from file neither writing to the output file as expected. It seems du.run_cmd is not useful when you need to interact with the cmd through stdin. I have tried setting stdout, stdin to None(keeping command as it is), but that didn't work either. Command seems to be waiting for input. Is their any problem if we use os.system().

Reply:
@dilip-krishnan Have you tried as_script=True argument to run_cmd?

Reply:
@hirenvadalia 
No, i haven't tried with as_script=True, are you sure that it will work with this parameter.

Reply:
@dilip-krishnan I think it will work.

Reply:
Ok thanks, will test it out then. Also is their any reason why we should not use os.system() in this particular case.

Reply:
Yes, mainly 2 reason:
1. As per document of os.system it says use of subprocess.Popen (which run_cmd using internally) is preferable then os.system
2. run_cmd provides you run command on remote host and many more controls over how your process should run.

Now after looking into your code, one question: What happen if you are running pbs_benchpress with -p "moms=\<remote mom which has alps_client command>" on system which does not have alps_client command? Where does alps_client command will be run if you use os.system()? on \<remote mom which has alps_client command>? or on system where you are running pbs_benchpress?